File: platforms/react-native/src/colliderGeometry.ts
/**
 * Pure geographic shape geometry for colliders and range circles. Each
 * function returns a closed ring of coordinates. NM → metres uses 1 NM = 1852 m.
 */

import { offset } from './geo.js';
import type { Coordinate } from './types.js';

const M_PER_NM = 1852;

/**
 * The 4 corners of a heading-aligned diamond (+ closing point):
 * front → headingDeg, right → +90°, back → +180°, left → +270°.
 */
export function diamond(
  center: Coordinate,
  forwardNM: number,
  sideNM: number,
  headingDeg: number
): Coordinate[] {
  const offsets: Array<[number, number]> = [
    [forwardNM * M_PER_NM, headingDeg], // front
    [sideNM * M_PER_NM, headingDeg + 90], // right
    [forwardNM * M_PER_NM, headingDeg + 180], // back
    [sideNM * M_PER_NM, headingDeg + 270], // left
  ];
  const pts = offsets.map(([d, b]) => offset(center, d, b));
  pts.push(pts[0]!); // close the shape
  return pts;
}

/** The 4 corners of a heading-aligned rectangle as [fL, fR, bR, bL, fL]. */
export function noseRect(
  center: Coordinate,
  forwardNM: number,
  sideNM: number,
  headingDeg: number
): Coordinate[] {
  const front = offset(center, forwardNM * M_PER_NM, headingDeg);
  const back = offset(center, forwardNM * M_PER_NM, headingDeg + 180);
  const fR = offset(front, sideNM * M_PER_NM, headingDeg + 90);
  const fL = offset(front, sideNM * M_PER_NM, headingDeg - 90);
  const bR = offset(back, sideNM * M_PER_NM, headingDeg + 90);
  const bL = offset(back, sideNM * M_PER_NM, headingDeg - 90);
  return [fL, fR, bR, bL, fL];
}

/** A geographic circle approximated as a polygon with `steps` segments. */
export function circle(
  center: Coordinate,
  radiusNM: number,
  steps = 36
): Coordinate[] {
  const pts: Coordinate[] = [];
  for (let i = 0; i < steps; i++) {
    pts.push(offset(center, radiusNM * M_PER_NM, (i * 360) / steps));
  }
  return pts;
}

/** Namespace object mirroring the Swift `ColliderGeometry` enum. */
export const ColliderGeometry = { diamond, noseRect, circle } as const;

File: packages/unity/com.zibaltech.geonavkit/Runtime/ColliderGeometry.cs
//
//  ColliderGeometry.cs
//  GeoNavKit
//
//  Pure geographic shape geometry for colliders and range circles. Each
//  method returns a closed ring of coordinates. NM → metres uses 1 NM = 1852 m.
//

namespace GeoNavKit
{
    public static class ColliderGeometry
    {
        private const double MetersPerNM = Distance.MetersPerNauticalMile;

        /// <summary>
        /// The 4 corners of a heading-aligned diamond (+ closing point):
        /// front → headingDeg, right → +90°, back → +180°, left → +270°.
        /// </summary>
        public static GeoCoordinate[] Diamond(GeoCoordinate center,
                                              double forwardNM, double sideNM,
                                              double headingDeg)
        {
            var pts = new GeoCoordinate[5];
            pts[0] = Geo.Offset(center, forwardNM * MetersPerNM, headingDeg);         // front
            pts[1] = Geo.Offset(center, sideNM * MetersPerNM, headingDeg + 90);       // right
            pts[2] = Geo.Offset(center, forwardNM * MetersPerNM, headingDeg + 180);   // back
            pts[3] = Geo.Offset(center, sideNM * MetersPerNM, headingDeg + 270);      // left
            pts[4] = pts[0];                                                          // close the shape
            return pts;
        }

        /// <summary>The 4 corners of a heading-aligned rectangle as [fL, fR, bR, bL, fL].</summary>
        public static GeoCoordinate[] NoseRect(GeoCoordinate center,
                                               double forwardNM, double sideNM,
                                               double headingDeg)
        {
            var front = Geo.Offset(center, forwardNM * MetersPerNM, headingDeg);
            var back = Geo.Offset(center, forwardNM * MetersPerNM, headingDeg + 180);
            var fR = Geo.Offset(front, sideNM * MetersPerNM, headingDeg + 90);
            var fL = Geo.Offset(front, sideNM * MetersPerNM, headingDeg - 90);
            var bR = Geo.Offset(back, sideNM * MetersPerNM, headingDeg + 90);
            var bL = Geo.Offset(back, sideNM * MetersPerNM, headingDeg - 90);
            return new[] { fL, fR, bR, bL, fL };
        }

        /// <summary>A geographic circle approximated as a polygon with <paramref name="steps"/> segments.</summary>
        public static GeoCoordinate[] Circle(GeoCoordinate center, double radiusNM, int steps = 36)
        {
            var pts = new GeoCoordinate[steps];
            for (int i = 0; i < steps; i++)
            {
                pts[i] = Geo.Offset(center, radiusNM * MetersPerNM, i * 360.0 / steps);
            }
            return pts;
        }
    }
}
